<a href="https://colab.research.google.com/github/q8611335/Human-activity-ML-model/blob/main/DataPrep.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Mount Google Drive

In [ ]:
# 最后导出code提交的时候可能要改一下这里
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Import dataset from Google Drive

In [ ]:
# file路径最后导出的时候可能要改
DATA_PATH = '/content/drive/MyDrive/COMP90049/LLCP2022.XPT'

# Read data

In [ ]:
import pandas as pd
import numpy as np

df_raw = pd.read_sas(DATA_PATH, format='xport', encoding='utf-8')
print(f"Data loading successful ")
print(f"Number of rows: {len(df_raw)}")
print(f"Number of columns:{df_raw.shape[1]}")

Data loading successful 
Number of rows: 445132
Number of columns:328


In [ ]:
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445132 entries, 0 to 445131
Columns: 328 entries, _STATE to _AIDTST4
dtypes: float64(323), object(5)
memory usage: 1.1+ GB


In [ ]:
# check variables we need
needed = {
    "filter":[
        "_SEX", "_STATE"
    ],
    "targets":[
        "MENTHLTH", "GENHLTH"
    ],
    "demographic":[
        "_AGE_G", "_RACE1", "EDUCA", "INCOME3",
        "MARITAL", "EMPLOY1", "CHILDREN"
    ],
    "behavioral":[
        "EXERANY2", "SLEPTIM1", "SMOKE100", "ALCDAY4", "_BMI5"
    ],
    "chronic":[
        "DIABETE4", "CVDCRHD4", "CVDSTRK3", "ASTHMA3",
        "CHCCOPD3", "HAVARTH4", "CHCKDNY2", "ADDEPEV3"
    ],
    "healthcare_access":[
        "PRIMINSR", "PERSDOC3", "MEDCOST1", "CHECKUP1"
    ],
    "ACE": [
        "ACEDEPRS", "ACEDRINK", "ACEDRUGS", "ACEPRISN", "ACEDIVRC",
        "ACEPUNCH", "ACEHURT1", "ACESWEAR", "ACETOUCH", "ACETTHEM",
        "ACEHVSEX", "ACEADSAF", "ACEADNED"
    ],

    "SDHE": [
        "LSATISFY", "EMTSUPRT", "SDHISOLT", "SDHEMPLY", "FOODSTMP",
        "SDHFOOD1", "SDHBILLS", "SDHUTILS", "SDHTRNSP", "SDHSTRE1"
    ]
}

all_needed_cols = []
for group, cols in needed.items():
    all_needed_cols.extend(cols)

all_needed_cols = list(dict.fromkeys(all_needed_cols))

print("Total needed variables:", len(all_needed_cols))
print(all_needed_cols)

existing_cols = [col for col in all_needed_cols if col in df_raw.columns]
print("Existing variables:", len(existing_cols))
print(existing_cols)

missing_cols = [col for col in all_needed_cols if col not in df_raw.columns]
print("\nMissing variables:", len(missing_cols))
print(missing_cols)


Total needed variables: 51
['_SEX', '_STATE', 'MENTHLTH', 'GENHLTH', '_AGE_G', '_RACE1', 'EDUCA', 'INCOME3', 'MARITAL', 'EMPLOY1', 'CHILDREN', 'EXERANY2', 'SLEPTIM1', 'SMOKE100', 'ALCDAY4', '_BMI5', 'DIABETE4', 'CVDCRHD4', 'CVDSTRK3', 'ASTHMA3', 'CHCCOPD3', 'HAVARTH4', 'CHCKDNY2', 'ADDEPEV3', 'PRIMINSR', 'PERSDOC3', 'MEDCOST1', 'CHECKUP1', 'ACEDEPRS', 'ACEDRINK', 'ACEDRUGS', 'ACEPRISN', 'ACEDIVRC', 'ACEPUNCH', 'ACEHURT1', 'ACESWEAR', 'ACETOUCH', 'ACETTHEM', 'ACEHVSEX', 'ACEADSAF', 'ACEADNED', 'LSATISFY', 'EMTSUPRT', 'SDHISOLT', 'SDHEMPLY', 'FOODSTMP', 'SDHFOOD1', 'SDHBILLS', 'SDHUTILS', 'SDHTRNSP', 'SDHSTRE1']
Existing variables: 51
['_SEX', '_STATE', 'MENTHLTH', 'GENHLTH', '_AGE_G', '_RACE1', 'EDUCA', 'INCOME3', 'MARITAL', 'EMPLOY1', 'CHILDREN', 'EXERANY2', 'SLEPTIM1', 'SMOKE100', 'ALCDAY4', '_BMI5', 'DIABETE4', 'CVDCRHD4', 'CVDSTRK3', 'ASTHMA3', 'CHCCOPD3', 'HAVARTH4', 'CHCKDNY2', 'ADDEPEV3', 'PRIMINSR', 'PERSDOC3', 'MEDCOST1', 'CHECKUP1', 'ACEDEPRS', 'ACEDRINK', 'ACEDRUGS', 'ACEPRIS

# Retrieve the required variables and inspect their properties

In [ ]:
# take our needed 51 cols
df_selected = df_raw[existing_cols].copy()
print("Selected data shape:", df_selected.shape)
display(df_selected.head())

Selected data shape: (445132, 51)


,_SEX,_STATE,MENTHLTH,GENHLTH,_AGE_G,_RACE1,EDUCA,INCOME3,MARITAL,EMPLOY1,...,LSATISFY,EMTSUPRT,SDHISOLT,SDHEMPLY,FOODSTMP,SDHFOOD1,SDHBILLS,SDHUTILS,SDHTRNSP,SDHSTRE1
0,2.0,1.0,88.0,2.0,6.0,1.0,6.0,99.0,1.0,7.0,...,1.0,1.0,5.0,2.0,2.0,5.0,2.0,2.0,2.0,4.0
1,2.0,1.0,88.0,1.0,6.0,1.0,4.0,5.0,3.0,2.0,...,1.0,1.0,5.0,2.0,2.0,5.0,2.0,2.0,2.0,5.0
2,2.0,1.0,3.0,2.0,5.0,1.0,6.0,10.0,1.0,7.0,...,2.0,2.0,3.0,2.0,2.0,5.0,2.0,2.0,2.0,5.0
3,2.0,1.0,88.0,1.0,6.0,1.0,4.0,77.0,1.0,7.0,...,1.0,1.0,3.0,2.0,2.0,5.0,2.0,2.0,2.0,5.0
4,2.0,1.0,88.0,4.0,3.0,1.0,5.0,5.0,1.0,5.0,...,1.0,1.0,5.0,2.0,2.0,5.0,2.0,2.0,2.0,5.0


In [ ]:
# generate variable checklist
check_rows = []
for group, cols in needed.items():
  for col in cols:
    check_rows.append({
        "group": group,
        "variable": col,
        "dtype": str(df_raw[col].dtype),
        "non_null_count": df_raw[col].notna().sum(),
        "raw_nan_count": df_raw[col].isna().sum(),
        "raw_nan_rate": df_raw[col].isna().mean(),
        "n_unique": df_raw[col].nunique(dropna=False)

    })

var_check_df = pd.DataFrame(check_rows)
display(var_check_df)

,group,variable,dtype,non_null_count,raw_nan_count,raw_nan_rate,n_unique
0,filter,_SEX,float64,445132,0,0.000000,2
1,filter,_STATE,float64,445132,0,0.000000,54
2,targets,MENTHLTH,float64,445129,3,0.000007,34
3,targets,GENHLTH,float64,445129,3,0.000007,8
4,demographic,_AGE_G,float64,445132,0,0.000000,6
5,demographic,_RACE1,float64,445130,2,0.000004,9
6,demographic,EDUCA,float64,445127,5,0.000011,8
7,demographic,INCOME3,float64,432200,12932,0.029052,14
8,demographic,MARITAL,float64,445124,8,0.000018,8
9,demographic,EMPLOY1,float64,438936,6196,0.013919,10


The variable checking table shows that all 51 key variables planned for this project are present in the BRFSS 2022 dataset, indicating that the proposed feature set is compatible with the raw data structure. The filter variables, target variables, demographic variables, chronic condition variables, and healthcare access variables are highly complete, with most raw missing rates close to 0%, making them suitable for the RQ1 baseline analysis. Some behavioural variables, such as SMOKE100, ALCDAY4, and _BMI5, show moderate missingness of approximately 7%–11%, which can be handled later through imputation. In contrast, the ACE module has very high missingness, around 89%, suggesting that ACE questions were only collected in a subset of states or respondents. The SDHE module also has substantial missingness, around 43%, indicating that it is not available for all respondents. Therefore, RQ1 can use a large female sample, while RQ2 and RQ3 will require additional filtering to retain women with valid ACE and SDHE information. It is also important to note that raw_nan_rate only captures blank values in the raw data and does not yet account for BRFSS special missing codes such as 7, 9, 77, and 99, which must be cleaned in the next preprocessing step.

# Data Cleaning - preprocessing

In [ ]:
# keep 51 columns needed
df_selected = df_raw[all_needed_cols].copy()
print("Selected data shape:", df_selected.shape)

Selected data shape: (445132, 51)


## Filter Female Data

In [ ]:
# filter female data
df_female = df_selected[df_selected["_SEX"] == 2].copy()
print("Female sample size::", df_female.shape)

Female sample size:: (235894, 51)


## Basic cleaning

In [ ]:
# Convert invalid codes
df_clean = df_female.copy()
invalid_codes = {
    # targets
    "MENTHLTH":[77, 99],
    "GENHLTH":[7, 9],

    # demographic
    "_RACE1": [9],
    "EDUCA": [9],
    "INCOME3": [77, 99],
    "MARITAL": [9],
    "EMPLOY1": [9],
    "CHILDREN": [99],

    # behavioral
    "EXERANY2": [7, 9],
    "SLEPTIM1": [77, 99],
    "SMOKE100": [7, 9],
    "ALCDAY4": [777, 999],
    "_BMI5": [9999],

    # chronic
    "DIABETE4": [7, 9],
    "CVDCRHD4": [7, 9],
    "CVDSTRK3": [7, 9],
    "ASTHMA3": [7, 9],
    "CHCCOPD3": [7, 9],
    "HAVARTH4": [7, 9],
    "CHCKDNY2": [7, 9],
    "ADDEPEV3": [7, 9],

    # healthcare access
    "PRIMINSR": [77, 99],
    "PERSDOC3": [7, 9],
    "MEDCOST1": [7, 9],
    "CHECKUP1": [7, 9],

}

# ACE
for col in needed["ACE"]:
  invalid_codes[col] = [7, 9]

# SDHE
for col in needed["SDHE"]:
  invalid_codes[col] = [7, 9]

# conversion
for col, codes in invalid_codes.items():
  if col in df_clean.columns:
    df_clean[col] = df_clean[col].replace(codes, np.nan)

print("Invalid code conversion completed.")
print("df_clean shape:", df_clean.shape)

Invalid code conversion completed.
df_clean shape: (235894, 51)


## Deal with special valid code

In [ ]:
# MENTHLTH: 88 - indicates that there were 0 days in the past 30 days with poor mental health.
df_clean["MENTHLTH"] = df_clean["MENTHLTH"].replace(88, 0)

# CHILDREN: 88 - have no kids
df_clean["CHILDREN"] = df_clean["CHILDREN"].replace(88, 0)

# _BMI5 = BMI * 100
# convert to real bmi value
df_clean["BMI"] = df_clean["_BMI5"] / 100

print("Special value conversion completed.")
display(df_clean[["MENTHLTH", "CHILDREN", "_BMI5", "BMI"]].head())

Special value conversion completed.


,MENTHLTH,CHILDREN,_BMI5,BMI
0,0.0,0.0,NaN,NaN
1,0.0,0.0,2657.0,26.57
2,3.0,0.0,2561.0,25.61
3,0.0,0.0,2330.0,23.30
4,0.0,0.0,2177.0,21.77


## Convert ALCDAY

In [ ]:
def convert_alcday4(x):
  if pd.isna(x):
    return np.nan

  # 888 = No alcohol consumption in the past 30 days
  if x == 888:
    return 0

  # 101-107 = how many days have drink per week
  # eg. 101 = drink 1 day in the past 30 days
  if 101 <= x <= 107:
    return (x-100) * 30 / 7

  # 201-230
  # eg. 205 = drink 5 days in the past 30 days
  if 201 <= x <= 230:
    return x - 200

  return np.nan

df_clean["alcohol_days_30"] = df_clean["ALCDAY4"].apply(convert_alcday4)

print("ALCDAY4 converted to alcohol_days_30.")
print(df_clean["alcohol_days_30"].describe())

ALCDAY4 converted to alcohol_days_30.
count    210816.000000
mean          4.047784
std           7.361696
min           0.000000
25%           0.000000
50%           0.000000
75%           4.285714
max          30.000000
Name: alcohol_days_30, dtype: float64


# build target feature

In [ ]:
# For RQ1 & RQ2
# target_fmd = 1 - frequent mental distress, MENTHLTH >= 14
df_clean["target_fmd"] = np.where(
    df_clean["MENTHLTH"] >= 14,
    1,
    np.where(df_clean["MENTHLTH"].notna(), 0, np.nan)
)

print("target_fmd distribution:")
print(df_clean["target_fmd"].value_counts(dropna=False))
print(df_clean["target_fmd"].value_counts(normalize=True, dropna=False))


target_fmd distribution:
target_fmd
0.0    194141
1.0     36872
NaN      4881
Name: count, dtype: int64
target_fmd
0.0    0.823001
1.0    0.156307
NaN    0.020691
Name: proportion, dtype: float64


In [ ]:
# For RQ3
# target_resilience = 1 - general health condiition is good (GENHLTH in 1, 2, 3)
# target_resilience = 0 - general health condition is poor (GENHLTH in 4, 5)
df_clean["target_resilience"] = np.where(
    df_clean["GENHLTH"].isin([1, 2, 3]),
    1,
    np.where(df_clean["GENHLTH"].isin([4, 5]), 0, np.nan)
)

print("\ntarget_resilience distribution:")
print(df_clean["target_resilience"].value_counts(dropna=False))
print(df_clean["target_resilience"].value_counts(normalize=True, dropna=False))


target_resilience distribution:
target_resilience
1.0    191414
0.0     43868
NaN       612
Name: count, dtype: int64
target_resilience
1.0    0.811441
0.0    0.185965
NaN    0.002594
Name: proportion, dtype: float64


## Build chronic count

In [ ]:
# 1 = Yes, 2 = No
chronic_yesno = [
    "CVDCRHD4", "CVDSTRK3", "ASTHMA3",
    "CHCCOPD3", "HAVARTH4", "CHCKDNY2", "ADDEPEV3"
]

for col in chronic_yesno:
  if col in df_clean.columns:
    df_clean[col + "_bin"] = df_clean[col].map({
        1: 1,
        2: 0
    })

# For DIABETE4 1 = yes, 0 = no
# 1 = Yes
# 2 = Yes, but only during pregnancy
# 3 = No
# 4 = No, pre-diabetes / borderline diabetes
if "DIABETE4" in df_clean.columns:
  df_clean["DIABETE4_bin"] = df_clean["DIABETE4"].map({
      1: 1,
      2: 0,
      3: 0,
      4: 0
  })

chronic_bin_cols = [
    col + "_bin" for col in needed["chronic"]
    if col + "_bin" in df_clean.columns
]

df_clean["chronic_count"] = df_clean[chronic_bin_cols].sum(axis=1, min_count=1)
print("chronic_count created.")
print("Used columns:", chronic_bin_cols)
print(df_clean["chronic_count"].describe())
print(df_clean["chronic_count"].value_counts(dropna=False).sort_index())

chronic_count created.
Used columns: ['DIABETE4_bin', 'CVDCRHD4_bin', 'CVDSTRK3_bin', 'ASTHMA3_bin', 'CHCCOPD3_bin', 'HAVARTH4_bin', 'CHCKDNY2_bin', 'ADDEPEV3_bin']
count    235792.000000
mean          1.173763
std           1.245035
min           0.000000
25%           0.000000
50%           1.000000
75%           2.000000
max           8.000000
Name: chronic_count, dtype: float64
chronic_count
0.0    86045
1.0    74815
2.0    42229
3.0    19576
4.0     8566
5.0     3200
6.0     1059
7.0      271
8.0       31
NaN      102
Name: count, dtype: int64


The variable `chronic_count` was created by converting eight chronic condition variables into binary indicators and summing them for each respondent.

## Create ACE_score

In [ ]:
# create binary ACE value & ACE score
# 1 = yes, 2 = no
ace_yesno = [
    "ACEDEPRS", "ACEDRINK", "ACEDRUGS", "ACEPRISN"
]

for col in ace_yesno:
  if col in df_clean.columns:
    df_clean[col + "_bin"] = df_clean[col].map({
        1: 1,
        2: 0
    })

# ACEDIVRC
# 1 = yes, 2 = no, 8 = Parents not married
# consider 8 as no parental divorce/separation
if "ACEDIVRC" in df_clean.columns:
  df_clean["ACEDIVRC_bin"] = df_clean["ACEDIVRC"].map({
      1: 1,
      2: 0,
      8: 0
  })

# ACE variables of frequency
# 1 = never, 2 = once, 3 = more than once
ace_freq = [
    "ACEPUNCH", "ACEHURT1", "ACESWEAR", "ACETOUCH", "ACETTHEM", "ACEHVSEX"
]
for col in ace_freq:
  if col in df_clean.columns:
    df_clean[col + "_bin"] = df_clean[col].map({
        # 1 = yes, 0 = no
        1: 0,
        2: 1,
        3: 1
    })

# positive / care variables
# 1 = never
# 2 = a little of the time, 3 = some of the time, 4 = most of the time
# 5 = all of the times
# Less care is considered adversity
for col in ["ACEADSAF", "ACEADNED"]:
  if col in df_clean.columns:
    df_clean[col + "_bin"] = df_clean[col].map({
        1: 1,
        2: 1,
        3: 1,
        4: 0,
        5: 0
    })

# total num of ACE binary cols
ace_bin_cols = [
    col + "_bin" for col in needed["ACE"]
    if col + "_bin" in df_clean.columns
]
# Record how many ACE items each sample answered
df_clean["ACE_answered_count"] = df_clean[ace_bin_cols].notna().sum(axis=1)

# if all ace items are missing, set ACE_score as NaN
df_clean["ACE_score"] = df_clean[ace_bin_cols].sum(axis=1, min_count=1)

# high ace score: ACE_score >= 4
df_clean["ACE_high_4"] = np.where(
    df_clean["ACE_score"] >= 4,
    1,
    np.where(df_clean["ACE_score"].notna(), 0, np.nan)
)

print("ACE_score created.")
print("Used ACE binary columns:", ace_bin_cols)
print("\nACE_answered_count distribution:")
print(df_clean["ACE_answered_count"].value_counts(dropna=False).sort_index())

print("ACE_score summary:")
print("\nACE_score distribution:")
print(df_clean["ACE_score"].value_counts(dropna=False).sort_index())

print("\nACE_high_4 distribution:")
print(df_clean["ACE_high_4"].value_counts(dropna=False))
print(df_clean["ACE_high_4"].value_counts(normalize=True, dropna=False))

ACE_score created.
Used ACE binary columns: ['ACEDEPRS_bin', 'ACEDRINK_bin', 'ACEDRUGS_bin', 'ACEPRISN_bin', 'ACEDIVRC_bin', 'ACEPUNCH_bin', 'ACEHURT1_bin', 'ACESWEAR_bin', 'ACETOUCH_bin', 'ACETTHEM_bin', 'ACEHVSEX_bin', 'ACEADSAF_bin', 'ACEADNED_bin']

ACE_answered_count distribution:
ACE_answered_count
0     210366
1         33
2         37
3         25
4         19
5         34
6         38
7         62
8         72
9         95
10       275
11       292
12      1389
13     23157
Name: count, dtype: int64
ACE_score summary:

ACE_score distribution:
ACE_score
0.0       9029
1.0       5189
2.0       3158
3.0       2148
4.0       1540
5.0       1192
6.0        894
7.0        706
8.0        541
9.0        432
10.0       289
11.0       209
12.0       142
13.0        59
NaN     210366
Name: count, dtype: int64

ACE_high_4 distribution:
ACE_high_4
NaN    210366
0.0     19524
1.0      6004
Name: count, dtype: int64
ACE_high_4
NaN    0.891782
0.0    0.082766
1.0    0.025452
Name: proportion,

ACE_score was constructed by recoding 13 ACE items into binary adversity indicators and summing them for each respondent. Respondents with no valid ACE item responses were assigned missing ACE_score rather than 0, because absence of module data does not imply absence of childhood adversity. Among female respondents, 25,528 had a valid ACE_score, and 6,004 were classified as high ACE using the threshold ACE_score >= 4.

Q: Why 4?

A: https://www.cdc.gov/violenceprevention/aces/about.html
Following the original CDC-Kaiser ACE Study literature, high childhood adversity was defined as ACE_score >= 4. The CDC describes the ACE score as “the total sum of the different categories of ACEs reported by participants” and reports ACE score prevalence using a “4 or more” category. Felitti et al. also specifically analysed “persons who had experienced four or more categories of childhood exposure”, showing that this group had substantially increased adult health risks compared with those with no ACE exposure.

## Create SDHE burden

In [ ]:
# create SDHE burden: indicates the quantity of current social needs/living difficulties

# LSATISFY: life satisfaction
# 1 = very satisfied, 2 = satisfied
# 3 = dissatisfied, 4 = very dissatisfied
# 3 & 4 are considered burden
df_clean["LSATISFY_burden"] = df_clean["LSATISFY"].map({
    1: 0,
    2: 0,
    3: 1,
    4: 1
})

# EMTSUPRT: Emotional support
# 1 = Always, 2 = Usually
# 3 = Sometimes, 4 = Rarely, 5 = Never
# 3, 4, 5 considered as low support (burden)
df_clean["EMTSUPRT_burden"] = df_clean["EMTSUPRT"].map({
    1: 0,
    2: 0,
    3: 1,
    4: 1,
    5: 1
})

# SDHISOLT: Socail isolation
# 1 = Always, 2 = Usually, 3 = Sometimes
# 4 = rarely, 5 = never
# 1, 2, 3 considered isolation (burden)
df_clean["SDHISOLT_burden"] = df_clean["SDHISOLT"].map({
    1: 1,
    2: 1,
    3: 1,
    4: 0,
    5: 0
})

# yes / no type SDHE variables
# 1 = yes, 2 = no
# indicates the respondent has experienced this social difficulty
# 1 - burden

yes_no_sdhe = [
    "SDHEMPLY",   # emplyment difficulty
    "FOODSTMP",   # used as material hardship proxy
    "SDHBILLS",   # difficulty paying bills
    "SDHUTILS",   # utility shutoff threat
    "SDHTRNSP"    # transpotation barrier
]

for col in yes_no_sdhe:
  if col in df_clean.columns:
    df_clean[col + "_burden"] = df_clean[col].map({
        1: 1,
        2: 0
    })

# SDHFOOD1: food insecurity
# 1 = Always, 2 = Usually, 3 = Sometims
# 4 = rarely, 5 = never
# 1, 2, 3 considered as burden
df_clean["SDHFOOD1_burden"] = df_clean["SDHFOOD1"].map({
    1: 1,
    2: 1,
    3: 1,
    4: 0,
    5: 0
})

# SDHSTRE1: Stress
# 1 = Always, 2 = Usually, 3 = Sometims
# 4 = rarely, 5 = never
# 1, 2, 3 considered as burden
df_clean["SDHSTRE1_burden"] = df_clean["SDHSTRE1"].map({
    1: 1,
    2: 1,
    3: 1,
    4: 0,
    5: 0
})

sdhe_burden_cols = [
    "LSATISFY_burden", "EMTSUPRT_burden", "SDHISOLT_burden",
    "SDHEMPLY_burden", "FOODSTMP_burden", "SDHBILLS_burden",
    "SDHUTILS_burden", "SDHTRNSP_burden", "SDHFOOD1_burden",
    "SDHSTRE1_burden"
]
sdhe_burden_cols = [
    col for col in sdhe_burden_cols
    if col in df_clean.columns
]
print("Number of SDHE burden columns:", len(sdhe_burden_cols))
print("Used SDHE burden columns:", sdhe_burden_cols)

Number of SDHE burden columns: 10
Used SDHE burden columns: ['LSATISFY_burden', 'EMTSUPRT_burden', 'SDHISOLT_burden', 'SDHEMPLY_burden', 'FOODSTMP_burden', 'SDHBILLS_burden', 'SDHUTILS_burden', 'SDHTRNSP_burden', 'SDHFOOD1_burden', 'SDHSTRE1_burden']


In [ ]:
# Record how many SDHE items each sample answered
df_clean["SDHE_answered_count"] = df_clean[sdhe_burden_cols].notna().sum(axis=1)

# If all SDHE items are missing, SDHE_burden remains NaN
df_clean["SDHE_burden"] = df_clean[sdhe_burden_cols].sum(axis=1, min_count=1)
print("SDHE_burden created.")

print("\nSDHE_answered_count distribution:")
print(df_clean["SDHE_answered_count"].value_counts(dropna=False).sort_index())

print("\nSDHE_burden summary:")
print(df_clean["SDHE_burden"].describe())

print("\nSDHE_burden distribution:")
print(df_clean["SDHE_burden"].value_counts(dropna=False).sort_index())


SDHE_burden created.

SDHE_answered_count distribution:
SDHE_answered_count
0     100572
1        187
2        180
3        163
4        177
5        239
6        244
7        311
8        805
9       4977
10    128039
Name: count, dtype: int64

SDHE_burden summary:
count    135322.000000
mean          1.499195
std           1.835173
min           0.000000
25%           0.000000
50%           1.000000
75%           2.000000
max          10.000000
Name: SDHE_burden, dtype: float64

SDHE_burden distribution:
SDHE_burden
0.0      53378
1.0      32775
2.0      19077
3.0      11766
4.0       7212
5.0       4601
6.0       2978
7.0       1942
8.0       1096
9.0        406
10.0        91
NaN     100572
Name: count, dtype: int64


SDHE_burden was constructed by converting 10 SDHE items into binary burden indicators and summing them for each respondent. Respondents with no valid SDHE item responses were assigned missing SDHE_burden rather than 0, because absence of module data does not imply absence of social needs. Among female respondents, 135,322 had a valid SDHE_burden score. The score ranged from 0 to 10, with a mean of approximately 1.50.

## Create ACE x SDHE

In [ ]:
# create ACE x SDHE
df_clean["ACE_x_SDHE"] = df_clean["ACE_score"] * df_clean["SDHE_burden"]
print("ACE_x_SDHE created.")
print(df_clean["ACE_x_SDHE"].describe())
print(df_clean["ACE_x_SDHE"].value_counts(dropna=False).sort_index())

ACE_x_SDHE created.
count    10434.000000
mean         5.480353
std         12.176733
min          0.000000
25%          0.000000
50%          0.000000
75%          5.000000
max        120.000000
Name: ACE_x_SDHE, dtype: float64
ACE_x_SDHE
0.0        5730
1.0         582
2.0         632
3.0         391
4.0         423
          ...  
104.0         1
108.0         2
117.0         1
120.0         2
NaN      225460
Name: count, Length: 66, dtype: int64


The interaction term `ACE_x_SDHE` was created by multiplying `ACE_score` and `SDHE_burden`. It is only available for respondents with both valid ACE and SDHE module data. In the female sample, 10,434 respondents had a valid interaction value, reflecting the overlap between the two optional BRFSS modules. The interaction term ranges from 0 to 120 in the cleaned data.

Q: Why ACE x SDHE?

A:

1. Interaction analysis is a recognised modelling approach for examining whether the joint effect of two exposures differs from what would be expected based on their individual effects.(VanderWeele & Knol, interaction tutorial)
2. Oregon State University's open epidemiology textbook explains:
Effect modification, also known as interaction, indicates that the relationship between exposure and outcome differs at different levels of a third variable. (Foundations of Epidemiology, effect modification)
3. The CDC explains that ACEs affect lifelong health and opportunities, noting their association with health outcomes such as chronic disease, behavioral health, and depression in adulthood. The CDC also mentions that differences in ACE risk are related to social and economic circumstances.
4. The product term ACE_score × SDHE_burden was included as a baseline interaction feature. This does not assume that the true relationship is strictly multiplicative; rather, it follows standard modelling practice for representing potential interaction or effect modification between two continuous predictors. More complex nonlinear relationships may be captured by tree-based or neural models, and the usefulness of the interaction term is evaluated empirically through model comparison.



# RQ1 dataset

In [ ]:
# RQ1 uses traditional baseline features only
# It should not be filtered by ACE or SDHE availability

rq1_features = (
    needed["demographic"] + ["EXERANY2", "SLEPTIM1", "SMOKE100", "alcohol_days_30", "BMI"]
    + needed["chronic"] + ["chronic_count"] + needed["healthcare_access"]
)

rq1_features = list(dict.fromkeys([col for col in rq1_features if col in df_clean.columns]))
df_rq1 = df_clean[
    df_clean["target_fmd"].notna()
][rq1_features + ["target_fmd"]].copy()

print("RQ1 dataset created.")
print("RQ1 dataset shape:", df_rq1.shape)

print("\nRQ1 target distribution:")
print(df_rq1["target_fmd"].value_counts(dropna=False))
print(df_rq1["target_fmd"].value_counts(normalize=True, dropna=False))

RQ1 dataset created.
RQ1 dataset shape: (231013, 26)

RQ1 target distribution:
target_fmd
0.0    194141
1.0     36872
Name: count, dtype: int64
target_fmd
0.0    0.84039
1.0    0.15961
Name: proportion, dtype: float64


RQ1 is a binary classification task with imbalanced classes. When modeling, don't just look at the accuracy, because a model that consistently predicts 0 can still achieve high accuracy.

We can use:
F1-score
Recall for class 1
ROC-AUC
PR-AUC
Balanced accuracy

For Logistic Regression/ Random Forest, we can use `class_weight = "balanced" `

# RQ2 Dataset

In [ ]:
# RQ2 uses baseline + ACE + SDHE features
# requires valid target_fmd, ACE_score, SDHE_burden

rq2_features = (
    rq1_features + needed["ACE"] + needed["SDHE"]
    + ["ACE_score", "SDHE_burden", "ACE_x_SDHE"]
)

rq2_features = list(dict.fromkeys([col for col in rq2_features if col in df_clean.columns]))
df_rq2 = df_clean[
    df_clean["target_fmd"].notna()
    & df_clean["ACE_score"].notna()
    & df_clean["SDHE_burden"].notna()
][rq2_features + ["target_fmd"]].copy()

print("RQ2 dataset created.")
print("RQ2 dataset shape:", df_rq2.shape)

print("\nRQ2 target distribution:")
print(df_rq2["target_fmd"].value_counts(dropna=False))
print(df_rq2["target_fmd"].value_counts(normalize=True, dropna=False))

RQ2 dataset created.
RQ2 dataset shape: (10264, 52)

RQ2 target distribution:
target_fmd
0.0    8675
1.0    1589
Name: count, dtype: int64
target_fmd
0.0    0.845187
1.0    0.154813
Name: proportion, dtype: float64


# RQ3 Dataset

In [ ]:
# RQ3 focuses on high-ACE (>=4) women and predicts resilience
# require valid SDHE_burden

rq3_features = (
    needed["SDHE"]
    + ["SDHE_burden"]
    + needed["healthcare_access"]
    + ["EXERANY2", "SLEPTIM1", "SMOKE100", "alcohol_days_30", "BMI"]
    + needed['demographic'] + ["chronic_count"]
)

rq3_features = list(dict.fromkeys([col for col in rq3_features if col in df_clean.columns]))
df_rq3 = df_clean[
    (df_clean["ACE_score"] >= 4) & (df_clean["target_resilience"].notna())
][rq3_features + ["target_resilience", "ACE_score"]].copy()

print("RQ3 dataset created.")
print("RQ3 dataset shape:", df_rq3.shape)
print("\nRQ3 target distribution:")
print(df_rq3["target_resilience"].value_counts(dropna=False))
print(df_rq3["target_resilience"].value_counts(normalize=True, dropna=False))


RQ3 dataset created.
RQ3 dataset shape: (5986, 30)

RQ3 target distribution:
target_resilience
1.0    4299
0.0    1687
Name: count, dtype: int64
target_resilience
1.0    0.718176
0.0    0.281824
Name: proportion, dtype: float64


# Save the datasets

In [ ]:
import os

SAVE_DIR = "/content/drive/MyDrive/COMP90049/processed"
os.makedirs(SAVE_DIR, exist_ok=True)

In [ ]:
df_clean.to_csv(f"{SAVE_DIR}/df_clean_female.csv", index=False)
df_rq1.to_csv(f"{SAVE_DIR}/df_rq1_baseline.csv", index=False)
df_rq2.to_csv(f"{SAVE_DIR}/df_rq2_augmentation.csv", index=False)
df_rq3.to_csv(f"{SAVE_DIR}/df_rq3_resilience.csv", index=False)

print("Processed datasets saved successfully.")

Processed datasets saved successfully.
